Add fallback check for when no contributions are highlighted in ArticleViewer

Abishekcs opened the pull request:
resolves #5461 


## What this PR does
* It Adds a fallback check for when no contributions are highlighted in ArticleViewer.
* **Bug-Fix** - It also Fixes a `CSS bug` in the ArticleViewer  when a `popover` is supposed to be shown for when the contributions made by the editor is not  `highlighted `  or  when the editor doesn't have a extant wikitext contributions in the `wikitext metadata` of the current version.


## Changes Made
- **article_viewer.jsx :**
 a) The primary changes in this file are related to the addition of the `fallback mechanism` for users whose contributions could not be highlighted. 
 - **article_viewer_api.js :**
    a) In this file, the method `fetchWikitextMetaData` has been added to the ArticleViewerAPI class. This method is responsible for fetching the wikitext metadata.
   
- **article_viewer_legend.jsx** :
     a) The unhighlightedContributors prop has been added to the ArticleViewerLegend component.
     b) CSS has been added `popup-style` to fix the `Bug` related to the `popup`.


## Screenshots
Before:
a) Related to the Bug: 

[Before Popover Bugfix.webm](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/74126284/c39f16f5-1fea-4afa-91e3-32ce369390a9)



b) Before Adding a fallback check

[Before(Enhancement).webm](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/74126284/dca20267-a590-4cc1-ba90-4696d441b47f)



After:
a) After Bugfix

[After Bugfix.webm](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/74126284/235146c6-78a8-4dae-92d8-704002882b31)



b)After Adding a fallback check

[After-enhancement.webm](https://github.com/WikiEducationFoundation/WikiEduDashboard/assets/74126284/e42dde0b-9d99-4627-9409-8042fd947266)

ragesoss commented on app/assets/javascripts/components/common/ArticleViewer/containers/ArticleViewer.jsx:
```diff
@@ -170,11 +171,31 @@ const ArticleViewer = ({ showOnMount, users, showArticleFinder, showButtonLabel,
     setHighlightedHtml(html);
   };
 
+  const checkFallback = (user) => {
```
As currently implemented, it looks like this would be run multiple times if multiple users edited the article but had no highlighted contributions. The wikitext metadata should only need to be fetched once at most.

ragesoss replied:
This function name is kind of vague. Can you think of a more precise one? A comment explaining the purpose of this would also be useful.

ragesoss commented on app/assets/javascripts/components/common/article_viewer_legend.jsx:
```diff
@@ -25,23 +24,28 @@ const ArticleViewerLegend = ({ article, users, colors, status, allUsers, failure
         let res;
+        // The 'unhighlightedContributors' array keeps track of the userids of users whose contributions
+        // were not successfully highlighted in the article viewer.
+        const isUserHovered = unhighlightedContributors.find(x => x === user.userid);
```
It looks like this constant is actually keeping track of whether the user had any contributions attributed in the `unhighlightedContributors` array. I don't think `isUserHovered` is a good name for that.

Abishekcs replied:
Okay will change variable name

ragesoss commented on app/assets/javascripts/components/common/article_viewer_legend.jsx:
```diff
@@ -25,23 +24,28 @@ const ArticleViewerLegend = ({ article, users, colors, status, allUsers, failure
           res = <div key={`legend-${user.name}`} className={'user-legend'}><a href={userLink} title={realName} target="_blank">{user.name}</a></div>;
         } else if (user.activeRevision === true) {
           res = <div key={`legend-${user.name}`} className={`user-legend ${colors[i]}`}><a href={userLink} title={realName} target="_blank">{user.name}</a><img className="user-legend-hover" style={{ color: 'transparent' }} src="/assets/images/arrow.svg" alt="scroll to users revisions" width="30px" height="20px" onClick={() => Scroller.scrollTo(user.name, scrollBox)} /></div >;
+        } else if (isUserHovered) {
```
This will be false while the fallback request is being made, so during that request, a user will have the no_highlighting message even if they actually have contributions and are about to get the contributionsNotHighlighted message. Ideally, instead of that, it would wait until the fallback request is completed before adding either message. I'm okay with cutting that corner if it's hard to implement, though. Shouldn't be too much of a problem for users.

Abishekcs commented:
Changes:

*  Changed the  `checkFallback ` function  name to `usersContributionExists` and
* Made the `Wikitextmetadata `to be fetched only once for multiple users of the same article.
* Changed  `isUserHovered` variable name to `UnhighlightedContributions`.
* Made the `HighlightedMessage` appear after only loading  of the authorship data is ready  depending on the `unhighlightedEditors` and ` highlightedHtml` in the " footer component".  So, that only the  `contributionsNotHighlighted`  or `no_highlighting` message appears for the user.
